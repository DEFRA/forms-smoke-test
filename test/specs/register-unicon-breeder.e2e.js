import { browser, expect } from '@wdio/globals'
import { before } from 'mocha'
import path from 'path'

import nameEntryPage from '~/test/page-objects/name.page.js'
import emailPage from '~/test/page-objects/email.page.js'
import phoneNumberPage from '~/test/page-objects/phone-number.page.js'
import addressPage from '~/test/page-objects/address.page.js'
import certificateAddressPage from '~/test/page-objects/certificate-address.page.js'
import policyStartDatePage from '~/test/page-objects/policy-start-date.page.js'
import uploadFilePage from '~/test/page-objects/upload-file.page.js'
import selectNoOfUnicornsPage from '~/test/page-objects/select-no-of-unicorns.page.js'
import selectTypeOfUnicornsPage from '~/test/page-objects/select-type-of-unicorns.page.js'
import unicornsTextPage from '~/test/page-objects/unicorns-text.page.js'
import noOfUnicornsStaffPage from '~/test/page-objects/no-of-unicorns-staff.page.js'
import summaryPage from '~/test/page-objects/summary.page.js'

describe('Register unicorn breeder form - e2e', () => {
  before(async () => {
    await nameEntryPage.open()
  })

  it('should enter name', async () => {
    await expect(browser).toHaveTitle(`What's your name? - e2e form - GOV.UK`)
    await nameEntryPage.enterName.setValue('John Doe')
    await nameEntryPage.submitButton.click()

    await expect(browser).toHaveTitle(
      `What's your email address? - e2e form - GOV.UK`
    )
  })

  it('should enter email address', async () => {
    await emailPage.enterEmail.setValue('john.doe@example.co.uk')
    await emailPage.submitButton.click()
    await expect(browser).toHaveTitle(
      `What's your phone number? - e2e form - GOV.UK`
    )
  })

  it('should enter phone number', async () => {
    await phoneNumberPage.enterPhoneNumber.setValue('07779875')
    await phoneNumberPage.submitButton.click()

    await expect(browser).toHaveTitle(
      `What's your address? - e2e form - GOV.UK`
    )
  })

  it('should enter address', async () => {
    await Promise.all([
      addressPage.enterAddressLine1.setValue('1 High Street'),
      addressPage.enterTown.setValue('Townsville'),
      addressPage.enterPostcode.setValue('TS1 1ST')
    ])
    await addressPage.submitButton.click()

    await expect(browser).toHaveTitle(
      'Do you want your unicorn breeder certificate sent to this address? - e2e form - GOV.UK'
    )
  })

  it('should select Yes to use same address for certificate', async () => {
    await certificateAddressPage.selectYesOrNo.click()
    await certificateAddressPage.submitButton.click()

    await expect(browser).toHaveTitle(
      'When does your unicorn insurance policy start? - e2e form - GOV.UK'
    )
  })

  it('should enter policy start date', async () => {
    await Promise.all([
      policyStartDatePage.enterDay.setValue('01'),
      policyStartDatePage.enterMonth.setValue('02'),
      policyStartDatePage.enterYear.setValue('2024')
    ])

    await policyStartDatePage.submitButton.click()

    await expect(browser).toHaveTitle(
      'Upload your insurance certificate - e2e form - GOV.UK'
    )
  })

  it('should upload file', async () => {
    const filePath = path.resolve('./test/file/test-file.txt')
    await uploadFilePage.chooseFile.setValue(filePath)
    await uploadFilePage.uploadFile.click()

    let fileUploaded = true

    while (fileUploaded) {
      const inLink = uploadFilePage.refreshPage

      await inLink.click()
      const uploadedFileElem = uploadFilePage.checkForUploadedFile

      if (await uploadedFileElem.isExisting()) {
        fileUploaded = false
      }
    }

    const uploadedElem = uploadFilePage.checkForUploadedFile
    await expect(uploadedElem).toBeDisplayed()
    await uploadFilePage.submitButton.click()

    await expect(browser).toHaveTitle(
      `How many unicorns do you expect to breed each year? - e2e form - GOV.UK`
    )
  })

  it('should select how many unicorns', async () => {
    await selectNoOfUnicornsPage.select1to5.click()
    await selectNoOfUnicornsPage.submitButton.click()

    await expect(browser).toHaveTitle(
      'What type of unicorns will you breed? - e2e form - GOV.UK'
    )
  })

  it('should select type of unicorns', async () => {
    await selectTypeOfUnicornsPage.selectFlying.click()
    await selectTypeOfUnicornsPage.selectAquatic.click()
    await selectTypeOfUnicornsPage.submitButton.click()

    await expect(browser).toHaveTitle(
      'Where will you keep the unicorns? - e2e form - GOV.UK'
    )
  })

  it('should enter where you will keep unicorns', async () => {
    await unicornsTextPage.enterWhereYouKeepUnicorn.setValue(
      'unicorns are kept in the garden'
    )
    await unicornsTextPage.submitButton.click()

    await expect(browser).toHaveTitle(
      'How many members of staff will look after the unicorns? - e2e form - GOV.UK'
    )
  })

  it('should enter no of unicorns staff', async () => {
    await noOfUnicornsStaffPage.enterNoOfUnicornStaff.setValue('5')
    await noOfUnicornsStaffPage.submitButton.click()

    await expect(browser).toHaveTitle('Summary - e2e form - GOV.UK')
  })

  it('should check summary page and submit form', async () => {
    await summaryPage.summary.isExisting()
    await summaryPage.contactDetails.isExisting()

    await expect(summaryPage.mainContent).toMatchSnapshot()

    await summaryPage.submitButton.click()

    await expect(browser).toHaveTitle('Form submitted - e2e form - GOV.UK')
  })
})
