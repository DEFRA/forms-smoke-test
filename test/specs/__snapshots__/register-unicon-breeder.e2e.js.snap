// Snapshot v1

exports[`Register unicorn breeder form - e2e > should check summary page and submit form 1`] = `
"<dl class="govuk-summary-list">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Unicorn insurance policy start date</dt>
    <dd class="govuk-summary-list__value app-prose-scope">1 February 2024</dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/when-does-your-unicorn-insurance-policy-start?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Unicorn insurance policy start date</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Documents</dt>
    <dd class="govuk-summary-list__value app-prose-scope">Uploaded 1 file</dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/upload-your-insurance-certificate?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Documents</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Number of unicorns</dt>
    <dd class="govuk-summary-list__value app-prose-scope">1 to 5</dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/how-many-unicorns-do-you-expect-to-breed-each-year?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Number of unicorns</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Type of unicorn</dt>
    <dd class="govuk-summary-list__value app-prose-scope">
      <ul>
        <li>Flying</li>
        <li>Aquatic</li>
      </ul>
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/what-type-of-unicorns-will-you-breed?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Type of unicorn</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Where you keep the unicorn</dt>
    <dd class="govuk-summary-list__value app-prose-scope">unicorns are kept in the garden</dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/where-will-you-keep-the-unicorns?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Where you keep the unicorn</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">Number of staff</dt>
    <dd class="govuk-summary-list__value app-prose-scope">5</dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link govuk-link--no-visited-state" href="/preview/draft/e2e-form/how-many-members-of-staff-will-look-after-the-unicorns?returnUrl=%2Fpreview%2Fdraft%2Fe2e-form%2Fsummary">Change
        <span class="govuk-visually-hidden">Number of staff</span>
      </a>
    </dd>
  </div>
</dl>"
`;
